File: BlackHole/clients/nodejs/test/protocol.test.js
/**
 * Codec tests that need no server.
 *
 * BlackHole Messaging - Gravicode Studios, led by Kang Fadhil.
 */

import test from 'node:test';
import assert from 'node:assert/strict';

import {
  FIXED_HEADER_SIZE,
  Message,
  MessageFlags,
  MessageType,
  PREFIX_SIZE,
  ProtocolError,
  StreamDescriptor,
  decodeFrame,
  encodeFrame,
  messageTypeName,
  topicMatches,
} from '../src/index.js';

test('round trips every field', () => {
  const original = new Message(
    MessageType.RpcRequest,
    'sensor/tank-3/temperature',
    Buffer.from('28.4'),
    987654321,
    MessageFlags.NoReply,
  );

  const frame = encodeFrame(original);
  const parsed = decodeFrame(frame);

  assert.equal(parsed.consumed, frame.length);
  assert.equal(parsed.message.type, original.type);
  assert.equal(parsed.message.flags, original.flags);
  assert.equal(parsed.message.header, original.header);
  assert.equal(parsed.message.correlationId, original.correlationId);
  assert.deepEqual(parsed.message.payload, original.payload);
});

test('frame layout matches the specification', () => {
  const frame = encodeFrame(new Message(MessageType.Publish, 'ab', Buffer.from('xyz'), 7));

  assert.equal(frame.readInt32LE(0), FIXED_HEADER_SIZE + 2 + 3);
  assert.equal(frame.readUInt8(4), MessageType.Publish);
  assert.equal(frame.readUInt8(5), 0);
  assert.equal(frame.readUInt16LE(6), 2);
  assert.equal(Number(frame.readBigInt64LE(8)), 7);
  assert.equal(frame.toString('utf8', PREFIX_SIZE, PREFIX_SIZE + 2), 'ab');
  assert.equal(frame.toString('utf8', PREFIX_SIZE + 2), 'xyz');
});

test('handles an empty header and payload', () => {
  const { message } = decodeFrame(encodeFrame(new Message(MessageType.Ping)));
  assert.equal(message.type, MessageType.Ping);
  assert.equal(message.header, '');
  assert.equal(message.payload.length, 0);
});

test('returns null until the whole frame arrives', () => {
  const frame = encodeFrame(new Message(MessageType.Publish, 'topic', Buffer.from('body')));

  for (let prefix = 0; prefix < frame.length; prefix++) {
    assert.equal(decodeFrame(frame.subarray(0, prefix)), null, `prefix ${prefix}`);
  }
  assert.notEqual(decodeFrame(frame), null);
});

test('parses back-to-back frames', () => {
  const stream = Buffer.concat(
    Array.from({ length: 5 }, (_, i) =>
      encodeFrame(new Message(MessageType.Publish, `topic/${i}`, Buffer.from(String(i)))),
    ),
  );

  const seen = [];
  let offset = 0;
  for (;;) {
    const parsed = decodeFrame(stream, offset);
    if (!parsed) break;
    offset += parsed.consumed;
    seen.push(parsed.message.header);
  }

  assert.deepEqual(seen, ['topic/0', 'topic/1', 'topic/2', 'topic/3', 'topic/4']);
  assert.equal(offset, stream.length);
});

test('handles non-ASCII headers and payloads', () => {
  const header = 'suhu/tangki/derajat-°C';
  const body = '28,4 °C — αβγ — 日本語 — 🕳';

  const { message } = decodeFrame(
    encodeFrame(new Message(MessageType.Publish, header, Buffer.from(body, 'utf8'))),
  );

  assert.equal(message.header, header);
  assert.equal(message.text(), body);
});

test('rejects a frame longer than the limit', () => {
  const frame = encodeFrame(new Message(MessageType.Publish, 't', Buffer.alloc(4096)));
  assert.throws(() => decodeFrame(frame, 0, 128), ProtocolError);
});

test('rejects an impossible length prefix', () => {
  assert.throws(() => decodeFrame(Buffer.from([2, 0, 0, 0, 1, 2])), /out of sync/);
});

test('rejects an oversized header', () => {
  assert.throws(() => encodeFrame(new Message(MessageType.Publish, 'x'.repeat(70000))), /Header is/);
});

test('preserves an unknown message type', () => {
  // A newer peer may know message types this client does not; that is not a framing error.
  const frame = encodeFrame(new Message(MessageType.Publish, 't'));
  frame.writeUInt8(0x7e, 4);

  const { message } = decodeFrame(frame);
  assert.equal(message.type, 0x7e);
  assert.equal(messageTypeName(0x7e), 'Unknown(0x7E)');
});

test('a decoded payload survives the read buffer being reused', () => {
  const frame = encodeFrame(new Message(MessageType.Publish, 't', Buffer.from('keep me')));
  const { message } = decodeFrame(frame);

  frame.fill(0); // Simulate the socket buffer being overwritten.
  assert.equal(message.text(), 'keep me');
});

test('StreamDescriptor round trips', () => {
  const original = new StreamDescriptor('kalibrasi-2026.csv', 1_048_576, 'text/csv');
  const parsed = StreamDescriptor.decode(original.encode());

  assert.equal(parsed.name, original.name);
  assert.equal(parsed.totalLength, original.totalLength);
  assert.equal(parsed.contentType, original.contentType);
  assert.equal(parsed.hasLength, true);
});

test('StreamDescriptor handles an unknown length', () => {
  const original = new StreamDescriptor('live.log', -1, 'text/plain');
  const parsed = StreamDescriptor.decode(original.encode());

  assert.equal(parsed.totalLength, -1);
  assert.equal(parsed.hasLength, false);
  assert.equal(StreamDescriptor.decode(Buffer.alloc(0)).hasLength, false);
});

test('topic matching mirrors the broker', () => {
  const cases = [
    ['sensor/tank-3/temp', 'sensor/tank-3/temp', true],
    ['sensor/+/temp', 'sensor/tank-3/temp', true],
    ['sensor/+/temp', 'sensor/tank-3/humidity', false],
    ['sensor/+/temp', 'sensor/a/b/temp', false],
    ['sensor/#', 'sensor/tank-3/temp', true],
    ['sensor/#', 'sensor', false],
    ['#', 'anything/at/all', true],
    ['sensor/tank-3/temp', 'sensor/tank-3', false],
    ['sensor/tank-3', 'sensor/tank-3/temp', false],
    ['+/+/temp', 'sensor/tank-3/temp', true],
  ];

  for (const [filter, topic, expected] of cases) {
    assert.equal(topicMatches(filter, topic), expected, `${filter} vs ${topic}`);
  }
});

File: BlackHole/clients/go/blackhole/protocol_test.go
// BlackHole Messaging - Gravicode Studios, led by Kang Fadhil.
//
// Codec tests that need no server.

package blackhole

import (
	"bytes"
	"encoding/binary"
	"strings"
	"testing"
)

func TestRoundTripsEveryField(t *testing.T) {
	original := Message{
		Type:          TypeRPCRequest,
		Flags:         FlagNoReply,
		CorrelationID: 987654321,
		Header:        "sensor/tank-3/temperature",
		Payload:       []byte("28.4"),
	}

	frame, err := EncodeFrame(original)
	if err != nil {
		t.Fatalf("encode: %v", err)
	}

	parsed, consumed, err := DecodeFrame(frame, DefaultMaxFrameLength)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if consumed != len(frame) {
		t.Errorf("consumed %d bytes, frame is %d", consumed, len(frame))
	}
	if parsed.Type != original.Type || parsed.Flags != original.Flags {
		t.Errorf("type/flags: got %v/%v, want %v/%v", parsed.Type, parsed.Flags, original.Type, original.Flags)
	}
	if parsed.CorrelationID != original.CorrelationID {
		t.Errorf("correlation: got %d, want %d", parsed.CorrelationID, original.CorrelationID)
	}
	if parsed.Header != original.Header {
		t.Errorf("header: got %q, want %q", parsed.Header, original.Header)
	}
	if !bytes.Equal(parsed.Payload, original.Payload) {
		t.Errorf("payload: got %q, want %q", parsed.Payload, original.Payload)
	}
}

func TestFrameLayoutMatchesTheSpecification(t *testing.T) {
	frame, err := EncodeFrame(Message{Type: TypePublish, Header: "ab", Payload: []byte("xyz"), CorrelationID: 7})
	if err != nil {
		t.Fatalf("encode: %v", err)
	}

	if got := binary.LittleEndian.Uint32(frame); got != FixedHeaderSize+2+3 {
		t.Errorf("length prefix: got %d, want %d", got, FixedHeaderSize+2+3)
	}
	if frame[4] != uint8(TypePublish) {
		t.Errorf("type byte: got %d", frame[4])
	}
	if frame[5] != 0 {
		t.Errorf("flags byte: got %d", frame[5])
	}
	if got := binary.LittleEndian.Uint16(frame[6:]); got != 2 {
		t.Errorf("header length: got %d, want 2", got)
	}
	if got := int64(binary.LittleEndian.Uint64(frame[8:])); got != 7 {
		t.Errorf("correlation id: got %d, want 7", got)
	}
	if string(frame[PrefixSize:PrefixSize+2]) != "ab" {
		t.Errorf("header bytes: got %q", frame[PrefixSize:PrefixSize+2])
	}
	if string(frame[PrefixSize+2:]) != "xyz" {
		t.Errorf("payload bytes: got %q", frame[PrefixSize+2:])
	}
}

func TestHandlesEmptyHeaderAndPayload(t *testing.T) {
	frame, _ := EncodeFrame(Message{Type: TypePing})
	parsed, _, err := DecodeFrame(frame, DefaultMaxFrameLength)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if parsed.Type != TypePing || parsed.Header != "" || len(parsed.Payload) != 0 {
		t.Errorf("got %v", parsed)
	}
}

func TestReturnsNothingUntilTheWholeFrameArrives(t *testing.T) {
	frame, _ := EncodeFrame(Message{Type: TypePublish, Header: "topic", Payload: []byte("body")})

	for prefix := 0; prefix < len(frame); prefix++ {
		_, consumed, err := DecodeFrame(frame[:prefix], DefaultMaxFrameLength)
		if err != nil {
			t.Fatalf("prefix %d: unexpected error %v", prefix, err)
		}
		if consumed != 0 {
			t.Fatalf("prefix %d: consumed %d bytes from a partial frame", prefix, consumed)
		}
	}

	if _, consumed, _ := DecodeFrame(frame, DefaultMaxFrameLength); consumed != len(frame) {
		t.Errorf("complete frame: consumed %d, want %d", consumed, len(frame))
	}
}

func TestParsesBackToBackFrames(t *testing.T) {
	var stream []byte
	for i := 0; i < 5; i++ {
		frame, _ := EncodeFrame(Message{Type: TypePublish, Header: "topic/" + string(rune('0'+i))})
		stream = append(stream, frame...)
	}

	var seen []string
	offset := 0
	for {
		message, consumed, err := DecodeFrame(stream[offset:], DefaultMaxFrameLength)
		if err != nil {
			t.Fatalf("decode: %v", err)
		}
		if consumed == 0 {
			break
		}
		offset += consumed
		seen = append(seen, message.Header)
	}

	if len(seen) != 5 || offset != len(stream) {
		t.Errorf("parsed %d frames over %d/%d bytes", len(seen), offset, len(stream))
	}
}

func TestHandlesNonASCII(t *testing.T) {
	original := Message{
		Type:    TypePublish,
		Header:  "suhu/tangki/derajat-°C",
		Payload: []byte("28,4 °C — 日本語 — 🕳"),
	}

	frame, _ := EncodeFrame(original)
	parsed, _, err := DecodeFrame(frame, DefaultMaxFrameLength)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if parsed.Header != original.Header || !bytes.Equal(parsed.Payload, original.Payload) {
		t.Errorf("got header %q payload %q", parsed.Header, parsed.Payload)
	}
}

func TestRejectsAFrameLongerThanTheLimit(t *testing.T) {
	frame, _ := EncodeFrame(Message{Type: TypePublish, Header: "t", Payload: make([]byte, 4096)})
	if _, _, err := DecodeFrame(frame, 128); err == nil {
		t.Fatal("expected an error for an oversized frame")
	} else if !strings.Contains(err.Error(), "exceeds") {
		t.Errorf("unexpected message: %v", err)
	}
}

func TestRejectsAnImpossibleLengthPrefix(t *testing.T) {
	if _, _, err := DecodeFrame([]byte{2, 0, 0, 0, 1, 2}, DefaultMaxFrameLength); err == nil {
		t.Fatal("expected an error for an impossible length prefix")
	} else if !strings.Contains(err.Error(), "out of sync") {
		t.Errorf("unexpected message: %v", err)
	}
}

func TestRejectsAnOversizedHeader(t *testing.T) {
	if _, err := EncodeFrame(Message{Type: TypePublish, Header: strings.Repeat("x", 70000)}); err == nil {
		t.Fatal("expected an error for an oversized header")
	}
}

func TestPreservesAnUnknownMessageType(t *testing.T) {
	// A newer peer may know message types this client does not; that is not a framing error.
	frame, _ := EncodeFrame(Message{Type: TypePublish, Header: "t"})
	frame[4] = 0x7E

	parsed, _, err := DecodeFrame(frame, DefaultMaxFrameLength)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if parsed.Type != MessageType(0x7E) {
		t.Errorf("got type %v", parsed.Type)
	}
	if got := parsed.Type.String(); got != "Unknown(0x7E)" {
		t.Errorf("String(): got %q", got)
	}
}

func TestStreamDescriptorRoundTrips(t *testing.T) {
	original := StreamDescriptor{Name: "kalibrasi-2026.csv", TotalLength: 1048576, ContentType: "text/csv"}
	parsed := DecodeStreamDescriptor(original.Encode())

	if parsed != original {
		t.Errorf("got %+v, want %+v", parsed, original)
	}
	if !parsed.HasLength() {
		t.Error("expected a known length")
	}
}

func TestStreamDescriptorHandlesUnknownLength(t *testing.T) {
	original := StreamDescriptor{Name: "live.log", TotalLength: UnknownLength, ContentType: "text/plain"}
	if parsed := DecodeStreamDescriptor(original.Encode()); parsed != original {
		t.Errorf("got %+v, want %+v", parsed, original)
	}
	if DecodeStreamDescriptor(nil).HasLength() {
		t.Error("an empty payload should decode as an unknown length")
	}
}

func TestTopicMatching(t *testing.T) {
	cases := []struct {
		filter, topic string
		want          bool
	}{
		{"sensor/tank-3/temp", "sensor/tank-3/temp", true},
		{"sensor/+/temp", "sensor/tank-3/temp", true},
		{"sensor/+/temp", "sensor/tank-3/humidity", false},
		{"sensor/+/temp", "sensor/a/b/temp", false},
		{"sensor/#", "sensor/tank-3/temp", true},
		{"sensor/#", "sensor", false},
		{"#", "anything/at/all", true},
		{"sensor/tank-3/temp", "sensor/tank-3", false},
		{"sensor/tank-3", "sensor/tank-3/temp", false},
		{"+/+/temp", "sensor/tank-3/temp", true},
	}

	for _, c := range cases {
		if got := TopicMatches(c.filter, c.topic); got != c.want {
			t.Errorf("TopicMatches(%q, %q) = %v, want %v", c.filter, c.topic, got, c.want)
		}
	}
}

func BenchmarkEncodeFrame(b *testing.B) {
	message := Message{Type: TypePublish, Header: "sensor/tank-3/temperature", Payload: []byte("28.4")}
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if _, err := EncodeFrame(message); err != nil {
			b.Fatal(err)
		}
	}
}

func BenchmarkDecodeFrame(b *testing.B) {
	frame, _ := EncodeFrame(Message{Type: TypePublish, Header: "sensor/tank-3/temperature", Payload: []byte("28.4")})
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if _, _, err := DecodeFrame(frame, DefaultMaxFrameLength); err != nil {
			b.Fatal(err)
		}
	}
}
